File: deno/src/base/permit/permit.service.ts
import type {
  PermitModel,
  GetUsrPermits,
} from "/gen/types.ts";

/**
 * 根据当前用户获取权限列表
 */
export async function getUsrPermits(): Promise<GetUsrPermits[]> {
  const {
    getAuthModel,
  } = await import("/lib/auth/auth.dao.ts");
  
  const {
    findAllPermit,
  } = await import("/gen/base/permit/permit.dao.ts");
  
  const {
    findByIdUsr,
  } = await import("/gen/base/usr/usr.dao.ts");
  
  const {
    findByIdMenu,
  } = await import("/gen/base/menu/menu.dao.ts");
  
  const {
    findAllRole,
  } = await import("/gen/base/role/role.dao.ts");
  
  const authModel = await getAuthModel(false);
  if (!authModel) {
    return [ ];
  }
  
  const options = {
    is_debug: false,
  };
  
  const usr_id = authModel.id;
  const usrModel = await findByIdUsr(usr_id, options);
  if (!usrModel) {
    return [ ];
  }
  const role_ids = usrModel.role_ids;
  if (!role_ids || role_ids.length === 0) {
    return [ ];
  }
  const roleModels = await findAllRole(
    {
      ids: role_ids,
    },
    undefined,
    undefined,
    options,
  );
  const permit_ids: PermitId[] = [ ];
  for (const roleModel of roleModels) {
    const permit_ids2 = roleModel.permit_ids;
    if (!permit_ids2 || permit_ids2.length === 0) {
      continue;
    }
    for (const permit_id of permit_ids2) {
      if (permit_ids.includes(permit_id as PermitId)) {
        continue;
      }
      permit_ids.push(permit_id as PermitId);
    }
  }
  // 切分成多个批次查询
  const permit_idsArr: PermitId[][] = [ ];
  const batch_size = 100;
  const batch_count = Math.ceil(permit_ids.length / batch_size);
  for (let i = 0; i < batch_count; i++) {
    permit_idsArr.push(permit_ids.slice(i * batch_size, (i + 1) * batch_size));
  }
  const permitModels: PermitModel[] = [ ];
  for (const permit_ids of permit_idsArr) {
    const permitModels0 = await findAllPermit(
      {
        ids: permit_ids,
      },
      undefined,
      undefined,
      options,
    );
    permitModels.push(...permitModels0);
  }
  const menu_idMap = new Map<MenuId, string>();
  for (const permitModel of permitModels) {
    const menu_id: MenuId = permitModel.menu_id;
    if (menu_idMap.has(menu_id)) {
      continue;
    }
    if (!menu_id) {
      menu_idMap.set(menu_id, "");
      continue;
    }
    const menuModel = await findByIdMenu(menu_id, options);
    if (!menuModel) {
      menu_idMap.set(menu_id, "");
      continue;
    }
    const route_path = menuModel.route_path;
    menu_idMap.set(menu_id, route_path);
  }
  const permits: GetUsrPermits[] = permitModels.map((permitModel) => {
    const menu_id: MenuId = permitModel.menu_id;
    const route_path = menu_idMap.get(menu_id) || "";
    return {
      id: permitModel.id as PermitId,
      menu_id,
      route_path,
      code: permitModel.code,
      lbl: permitModel.lbl,
    };
  });
  return permits;
}

/**
 * 查看当前用户是否有权限
 */
export async function usePermit(
  route_path: string,
  code: string,
): Promise<void> {
  const {
    ns,
  } = await import("/src/base/i18n/i18n.ts");
  
  const {
    getAuthModel,
  } = await import("/lib/auth/auth.dao.ts");
  
  const {
    findOnePermit,
  } = await import("/gen/base/permit/permit.dao.ts");
  
  const {
    findByIdUsr,
  } = await import("/gen/base/usr/usr.dao.ts");
  
  const {
    findAllRole,
  } = await import("/gen/base/role/role.dao.ts");
  
  const {
    findOneMenu,
  } = await import("/gen/base/menu/menu.dao.ts");
  
  const options = {
    is_debug: false,
  };
  
  const menuModel = await findOneMenu(
    {
      route_path,
      is_enabled: [ 1 ],
    },
    undefined,
    options,
  );
  if (!menuModel) {
    return;
  }
  
  const authModel = await getAuthModel(false);
  
  if (!authModel) {
    throw await ns("无权限");
  }
  const usr_id: UsrId = authModel.id;
  const usrModel = await findByIdUsr(usr_id, options);
  if (!usrModel) {
    throw await ns("无权限");
  }
  if (usrModel.username === "admin") {
    return;
  }
  const role_ids = usrModel.role_ids;
  if (!role_ids || role_ids.length === 0) {
    throw await ns("无权限");
  }
  const roleModels = await findAllRole(
    {
      ids: role_ids,
    },
    undefined,
    undefined,
    options,
  );
  const permit_ids: PermitId[] = [ ];
  for (const roleModel of roleModels) {
    const permit_ids2 = roleModel.permit_ids;
    if (!permit_ids2 || permit_ids2.length === 0) {
      continue;
    }
    for (const permit_id of permit_ids2) {
      if (permit_ids.includes(permit_id)) {
        continue;
      }
      permit_ids.push(permit_id);
    }
  }
  // 切分成多个批次查询
  const permit_idsArr: PermitId[][] = [ ];
  const batch_size = 100;
  const batch_count = Math.ceil(permit_ids.length / batch_size);
  for (let i = 0; i < batch_count; i++) {
    permit_idsArr.push(permit_ids.slice(i * batch_size, (i + 1) * batch_size));
  }
  for (const permit_ids of permit_idsArr) {
    const permitModel = await findOnePermit(
      {
        ids: permit_ids,
        menu_id: [ menuModel.id ],
        code,
      },
      undefined,
      options,
    );
    if (permitModel) {
      return;
    }
  }
  throw await ns("{0} {1} 无权限", menuModel.lbl, code);
}

File: rust/src/common/permit/permit_service.rs
use std::collections::HashMap;

use color_eyre::eyre::{Result,eyre};

use crate::common::context::{
  get_auth_model,
  Options,
};

use crate::common::i18n::i18n_dao::ns;

use crate::r#gen::base::menu::menu_dao::{
  find_by_id_menu,
  find_one_menu,
};
use crate::r#gen::base::menu::menu_model::{MenuSearch, MenuId};

use super::permit_model::GetUsrPermits;

use crate::r#gen::base::usr::usr_dao::find_by_id_usr;

use crate::r#gen::base::role::role_dao::find_all_role;
use crate::r#gen::base::role::role_model::RoleSearch;

use crate::r#gen::base::permit::permit_dao::{
  find_all_permit,
  find_one_permit,
};
use crate::r#gen::base::permit::permit_model::PermitSearch;
use crate::r#gen::base::permit::permit_model::PermitModel;
use crate::r#gen::base::permit::permit_model::PermitId;

/// 根据当前用户获取权限列表
pub async fn get_usr_permits() -> Result<Vec<GetUsrPermits>> {
  let auth_model = get_auth_model();
  if auth_model.is_none() {
    return Ok(Vec::new());
  }
  let auth_model = auth_model.unwrap();
  
  let options = Options::new();
  let options = options.set_is_debug(Some(false));
  let options = Some(options);
  
  let usr_model = find_by_id_usr(
    auth_model.id,
    options.clone(),
  ).await?;
  if usr_model.is_none() {
    return Ok(Vec::new());
  }
  let usr_model = usr_model.unwrap();
  
  let role_ids = usr_model.role_ids;
  if role_ids.is_empty() {
    return Ok(Vec::new());
  }
  
  let role_models = find_all_role(
    RoleSearch {
      ids: role_ids.into(),
      is_enabled: vec![1].into(),
      ..Default::default()
    }.into(),
    None,
    None,
    options.clone(),
  ).await?;
  
  let mut permit_ids = Vec::<PermitId>::new();
  for role_model in role_models {
    for permit_id in role_model.permit_ids {
      if permit_ids.contains(&permit_id) {
        continue;
      }
      permit_ids.push(permit_id);
    }
  }
  let permit_len = permit_ids.len();
  
  // 切分成多个批次查询
  let batch_size = 100;
  let mut batch_count = permit_len / batch_size;
  if permit_len % batch_size != 0 {
    batch_count += 1;
  }
  let batch_count = batch_count;
  let mut permit_ids_arr = Vec::<Vec<PermitId>>::with_capacity(batch_count);
  for i in 0..batch_count {
    let start = i * batch_size;
    let mut end = (i + 1) * batch_size;
    if end > permit_len {
      end = permit_len;
    }
    let end = end;
    let permit_ids = permit_ids[start..end].to_vec();
    permit_ids_arr.push(permit_ids);
  }
  let permit_ids_arr = permit_ids_arr;
  
  let mut permit_models: Vec<PermitModel> = Vec::with_capacity(permit_len);
  for permit_ids in permit_ids_arr {
    let mut permit_models_tmp = find_all_permit(
      PermitSearch {
        ids: permit_ids.into(),
        ..Default::default()
      }.into(),
      None,
      None,
      options.clone(),
    ).await?;
    permit_models.append(&mut permit_models_tmp);
  }
  let permit_models = permit_models;
  
  let mut menu_id_map = HashMap::<MenuId, String>::with_capacity(permit_len);
  
  for permit_model in permit_models.iter() {
    let menu_id = permit_model.menu_id.clone();
    if menu_id_map.contains_key(&menu_id) {
      continue;
    }
    if menu_id.is_empty() {
      menu_id_map.insert(menu_id.clone(), "".to_owned());
    }
    
    let menu_model = find_by_id_menu(
      menu_id.clone(),
      options.clone(),
    ).await?;
    
    if menu_model.is_none() {
      menu_id_map.insert(menu_id.clone(), "".to_owned());
      continue;
    }
    let menu_model = menu_model.unwrap();
    
    let route_path = menu_model.route_path;
    menu_id_map.insert(menu_id, route_path);
  }
  
  let permits: Vec<GetUsrPermits> = permit_models.into_iter()
    .map(|item| {
      let menu_id = item.menu_id;
      let route_path: Option<&String> = menu_id_map.get(&menu_id);
      let route_path: String = route_path
        .map_or_else(
          || "".to_owned(),
          |item| item.to_owned()
        );
      GetUsrPermits {
        id: item.id,
        menu_id,
        route_path,
        code: item.code,
        lbl: item.lbl,
      }
    })
    .collect();
  
  Ok(permits)
}

/// 后端按钮权限校验
pub async fn use_permit(
  route_path: String,
  code: String,
) -> Result<()> {
  
  let options = Options::new()
    .set_is_debug(Some(false));
  let options = Some(options);
  
  let menu_model = find_one_menu(
    MenuSearch {
      route_path: route_path.clone().into(),
      is_enabled: vec![1].into(),
      ..Default::default()
    }.into(),
    None,
    options.clone(),
  ).await?;
  if menu_model.is_none() {
    return Ok(());
  }
  let menu_model = menu_model.unwrap();
  
  let auth_model = get_auth_model();
  
  if auth_model.is_none() {
    let err_msg = ns(
      "无权限".to_owned(),
      None,
    ).await?;
    return Err(eyre!(err_msg));
  }
  
  let auth_model = auth_model.unwrap();
  
  let usr_id = auth_model.id;
  
  let usr_model = find_by_id_usr(
    usr_id,
    options.clone(),
  ).await?;
  
  if usr_model.is_none() {
    let err_msg = ns(
      "无权限".to_owned(),
      None,
    ).await?;
    return Err(eyre!(err_msg));
  }
  
  let usr_model = usr_model.unwrap();
  
  if usr_model.username == "admin" {
    return Ok(());
  }
  
  let role_ids = usr_model.role_ids;
  
  if role_ids.is_empty() {
    let err_msg = ns(
      "无权限".to_owned(),
      None,
    ).await?;
    return Err(eyre!(err_msg));
  }
  
  let role_models = find_all_role(
    RoleSearch {
      ids: role_ids.into(),
      is_enabled: vec![1].into(),
      ..Default::default()
    }.into(),
    None,
    None,
    options.clone(),
  ).await?;
  
  // 过滤掉重复的 permit_ids
  let mut permit_ids = Vec::<PermitId>::new();
  for role_model in role_models.into_iter() {
    for permit_id in role_model.permit_ids.into_iter() {
      if permit_ids.contains(&permit_id) {
        continue;
      }
      permit_ids.push(permit_id);
    }
  }
  let permit_ids = permit_ids;
  
  // 切分成多个批次查询
  let mut permit_ids_arr = Vec::<Vec<PermitId>>::new();
  let batch_size = 100;
  let batch_count = (permit_ids.len() / batch_size) + 1;
  
  for i in 0..batch_count {
    let start = i * batch_size;
    let mut end = (i + 1) * batch_size;
    if end > permit_ids.len() {
      end = permit_ids.len();
    }
    let permit_ids = permit_ids[start..end].to_vec();
    permit_ids_arr.push(permit_ids);
  }
  let menu_id = menu_model.id;
  for permit_ids in permit_ids_arr.into_iter() {
  
    let permit_model = find_one_permit(
      PermitSearch {
        ids: permit_ids.into(),
        menu_id: vec![menu_id.clone()].into(),
        code: code.clone().into(),
        ..Default::default()
      }.into(),
      None,
      options.clone(),
    ).await?;
    
    if permit_model.is_some() {
      return Ok(());
    }
  }
  
  let mut map: HashMap<String, String> = HashMap::with_capacity(2);
  map.insert("0".to_owned(), menu_model.lbl);
  map.insert("1".to_owned(), code);
  
  let err_msg = ns(
    "{0} {1} 无权限".to_owned(),
    Some(map),
  ).await?;
  
  Err(eyre!(err_msg))
}
